File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = {i: [] for i in range(numCourses)}
        for course, pre in prerequisites:
            graph[pre].append(course)

        visit = [0] * numCourses 

        def dfs(course):
            if visit[course] == 1:
                return False  
            if visit[course] == 2:
                return True   
            
            visit[course] = 1  
            for neighbor in graph[course]:
                if not dfs(neighbor):
                    return False
            visit[course] = 2
            return True

        for c in range(numCourses):
            if not dfs(c):
                return False
        return True

File: 0207-course-schedule/0207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    let graph = Array.from({length:numCourses}, ()=>[])
    for (let [a,b] of prerequisites){
        graph[b].push(a)

    }
    let state = new Array(numCourses).fill(0)
    function dfs(course){
        if(state[course]===1)return false
        if(state[course]===2) return true

        state[course] = 1
        for(let next of graph[course]){
            if(!dfs(next)) return false
        }
        state[course] = 2
        return true
    }
    for(let i=0; i<numCourses; i++){
        if(!dfs(i)) return false
    }
    return true
};